### PR TITLE
add group-lines() to correlate multi-line messages and "smart" mode multi-line mode

### PR DESCRIPTION
**Pull request by bazsi**
This PR is a combination of two features plus some refactoring to make the whole thing possible:

`multi-line-mode(smart)`

    `multi-line-mode(smart)`: a new multi-line mode was added to recognize
    backtraces for various programming languages automatically.  With this
    multi-line mode, the inherently multi-line data backtrace format is
    recognized even if they span multiple lines in the input and are converted
    to a single log message for easier analysis.  Backtraces for the following
    programming languages are recognized : Python, Java, JavaScript, PHP, Go,
    Ruby and Dart.
    
    The regular expressions to recognize these programming languages are
    specified by an external file called
    `/usr/share/syslog-ng/smart-multi-line.fsm` (installation path depends on
    configure arguments), in a format that is described in that file.
    
`group-lines()` parser

    `group-lines()` parser: this new parser correlates multi-line messages
    received as separate, but subsequent lines into a single log message.
    Received messages are first collected into streams related messages (using
    key()), then collected into correlation contexts up to timeout() seconds.
    The identification of multi-line messages are then performed on these
    message contexts within the time period.
    
    ```
      group-lines(key("$FILE_NAME")
                  multi-line-mode("smart")
	          template("$MESSAGE")
	          timeout(10)
	          line-separator("\n")
      );
    ```
    
    Configuration options are as follows:
      key() -- specifies a template that determines what messages are forming a
               single stream. Messages where the template expansion results in the same
               key are considered part of the same stream. Using key() you can apply
               multi-line extraction even if different streams are interleaved in your
               input.
    
      multi-line-mode() -- same as multi-line-mode() for multi-line file
                           sources.  Value is one of: `smart`, `indented`,
                           `prefix-garbage` or `prefix-suffix`.
    
      multi-line-prefix(), multi-line-suffix(), multi-line-garbage() --
                           regular expressions associated with regexp based multi-line-mode
                           values.
    
      template() -- a template string that specifies what constitutes an line to
                    group-lines(). In simplest cases this is $MSG or $RAWMSG.
    
      line-separator() -- in case a multi-line message is found, this value is
                          inserted between segments of the new multi-line message. Defaults to '\n'
                          that is the newline character.

### CODE REVIEW

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by bazsi** [comment]
this should now run green and is getting ready to review.

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by bazsi** [comment]
hmm.. the .tsv file is only read from the install location, and at make check that is not yet run. :(

going back to WIP

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by bazsi** [comment]
Todo: end-to-end test via light.

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by kira-syslogng** [comment]
Build FAILURE

**Review by alltilla — COMMENTED** [review]
Reviewed up to (including) `grouping-parser: reuse CorrelationContext instances (488027200)`. I only had one minor, optional comment, so I am open to merge the code up to that point. I think that commit would be a nice point to separate this PR.

Continuing with the rest of the commits.

File: modules/correlation/tests/test_patterndb.c
Comment (bazsi) [reply modules/correlation/tests/test_patterndb.c:123]: This should match the order in which syslog-ng itself is generating the messages: e.g. generate synthetic messages first and then the message that is under evaluation.

So actually I think this improves the similarities between the production and test environments.

I would leave this here.

Do you want me to submit the beginning of this branch as a separate PR?

Comment (alltilla) [reply modules/correlation/tests/test_patterndb.c:123]: Thanks for the explanation.

It would be nice if you could open a separate PR with those commits, yes.

File: lib/multi-line/smart-multi-line.c
Comment (alltilla) [inline lib/multi-line/smart-multi-line.c:301]: This function seemed a bit overcomplicated to me (might be a problem with me), so I have reordered it a bit. Things that did not fit for me:
  1. `segment_is_trace` seems to be the same as `*segment_is_part_of_trace`.
  2. There seems to be some `if` + `=`s, which can be simplified to be just `=`.
  3. I think we should allow a rule to start and end a trace at the same time. While it seems weird and I cannot think of any examples of this, I think we should let the user write optionally one-line traces, for example:
```
# dummy trace
start_state,dummy_trace	/dummy trace: optional line: .*/	dummy_trace
start_state,dummy_trace	/dummy trace: mandatory ending with simple conclusion: .*/	start_state
```

The code after my reordering:
```c
void
_process_segment(SmartMultiLine *self, const gchar *segment, gsize segment_len,
                 gboolean *segment_is_part_of_trace,
                 gboolean *segment_starts_a_new_trace,
                 gboolean *segment_ends_trace)
{
  gboolean processing_started_from_start_state = self->current_state == SMLS_START_STATE;

  *segment_is_part_of_trace = _fsm_transition(self, segment, segment_len);
  msg_trace_printf("smart-multi-line: [STEP1] >>%.*s<<, result=%d, state=%d", (int) segment_len, segment,
                   *segment_is_part_of_trace, self->current_state);

  if (*segment_is_part_of_trace)
    {
      *segment_starts_a_new_trace = processing_started_from_start_state;
    }
  else
    {
      /* try again from the start state, the current segment may be part of a new trace */
      *segment_is_part_of_trace = *segment_starts_a_new_trace = _fsm_transition(self, segment, segment_len);
      msg_trace_printf("smart-multi-line: [STEP2]: >>%.*s<<, result=%d, state=%d", (int) segment_len, segment,
                       *segment_is_part_of_trace, self->current_state);
    }

  *segment_ends_trace = *segment_is_part_of_trace && self->current_state == SMLS_START_STATE;
}
```

File: modules/correlation/group-lines.c
Comment (alltilla) [inline modules/correlation/group-lines.c:267]: [optional]

It seems weird to me, that we change the `clone_id` of `self` here. Same applies for `groupingby`.

File: modules/correlation/group-lines.c
Comment (bazsi) [reply modules/correlation/group-lines.c:2]: true that.

File: modules/correlation/group-lines.c
Comment (bazsi) [reply modules/correlation/group-lines.c:248]: added, thanks. this was moved by to grouping-by() during the refactorings to produce better warnings

File: modules/correlation/group-lines.c
Comment (bazsi) [reply modules/correlation/group-lines.c:292]: I think with grouping-by() is a more conscious choice. With a multi-line message, 10 seconds should be plenty to collect messages belonging to the same multi-line message. 

But I am ok with removing the default.

File: modules/correlation/group-lines.c
Comment (bazsi) [reply modules/correlation/group-lines.c:262]: did both of these.

File: modules/correlation/group-lines.c
Comment (bazsi) [reply modules/correlation/group-lines.c:267]: this is because we need to differentiate all clones as they will all use their own state.

File: modules/correlation/group-lines.h
Comment (bazsi) [reply modules/correlation/group-lines.h:23]: fixed up all headers.

File: modules/correlation/group-lines.c
Comment (alltilla) [reply modules/correlation/group-lines.c:267]: Yes, I agree, but I think we should not modify `self` with `self->clone_id++` in the clone fn. IMO `cloned->clone_id = self->clone_id + 1` would be better.

File: lib/multi-line/smart-multi-line.c
Comment (bazsi) [reply lib/multi-line/smart-multi-line.c:348]: since we return MLL_EXTRACTED here, no msg remains partially consumed. But I agree, it might be clearer, so I added an explicit statement to set consumed_message_is_trace to FALSE.

note that this is not going to be used, at least with the current code.

Also note, that consumed_message_is_trace refers to the partially consumed message while segment_is_part_of_trace refers only to the current segment and shows whether it is considered to be a continuation.

File: lib/multi-line/smart-multi-line.c
Comment (bazsi) [reply lib/multi-line/smart-multi-line.c:129]: removed.

File: lib/multi-line/smart-multi-line.c
Comment (bazsi) [reply lib/multi-line/smart-multi-line.c:257]: I don't understand this point.

File: modules/correlation/group-lines.c
Comment (bazsi) [reply modules/correlation/group-lines.c:279]: the issue is there might be multiple clones and each needs a different ID.

**Comment by bazsi** [comment]
I fixed most issues, except the refactor for the fsm_transition() function, which seems nicer, I'll just need to replay the refactoring steps myself to see if all corner cases are covered.

I wrote that function ~5 months ago and although I did improve it somewhat, you are right that it needs some more improvements, but let's nail the rest of the comments first.

File: modules/correlation/group-lines.c
Comment (alltilla) [reply modules/correlation/group-lines.c:279]: Thanks for the explanation, I understand now why it is done like this. I don't think we should change it in this PR, as groupingby already has this logic, but I think this is not the way to go in the long run. My problem is: The persist name of the parser is dependent of how many times it was cloned. Let's say we have a config, where it gets cloned 2 times. It has `clone_id=2` in its persist name and stores its values in the persist with that name. We add another reference/user of the parser in the config, and reload. Now it gets `clone_id=3`, and it won't find its persist entry from before reload and starts a new, however the new clone gets an already filled persist entry.

File: lib/multi-line/smart-multi-line.c
Comment (alltilla) [reply lib/multi-line/smart-multi-line.c:300]: We can leave this as is.

**Review by alltilla — APPROVED** [review]
LGTM, after the copyright check is fixed we can merge this.

**Comment by alltilla** [comment]
I have added the new files to the copyright policy.

**Review by alltilla — APPROVED** [review]

**Pull request merged**
